### PR TITLE
test: to check the status of pvc as bound and pod as running

### PR DESCRIPTION
**Pull request by riya-singhal31**
this test creates pvc and pod and verifies the status of pvc to be bound and pod to be running, and cleans them up.

Signed-off-by: riya-singhal31 <riyasinghalji@gmail.com>

### CODE REVIEW

**Comment by riya-singhal31** [comment]
Output:
[riyasinghal@fedora lvm-operator]$ oc get pvc,pods
NAME                                STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
persistentvolumeclaim/lvmblockpvc   Bound    pvc-8d8a4385-99c0-46b5-a0ba-235983d2aed9   5Gi        RWO            odf-lvm-vg1    10s
persistentvolumeclaim/lvmfilepvc    Bound    pvc-244885b9-7e9d-40b4-88cb-956bb040bdd6   5Gi        RWO            odf-lvm-vg1    12s

NAME              READY   STATUS    RESTARTS   AGE
pod/lvmblockpod   1/1     Running   0          10s
pod/lvmfilepod    1/1     Running   0          12s

File: e2e/common.go [hunk 1]
```diff
@@ -1,12 +1,128 @@
...
+// GetSamplePVC returns a sample pvc.
+func GetSamplePVC(storageClass string, quantity string, name string, volumemode v1.PersistentVolumeMode) *k8sv1.PersistentVolumeClaim {
+	storageQuantity, err := resource.ParseQuantity(quantity)
+	gomega.Expect(err).To(gomega.BeNil())
+
+	pvc := &k8sv1.PersistentVolumeClaim{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "PersistentVolumeClaim",
+		},
```
Comment (Madhu-1) [inline e2e/common.go:32]: You dont need to add the API version and kind

File: e2e/common.go [hunk 2]
```diff
@@ -1,12 +1,128 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: TestNamespace,
+		},
+		Spec: k8sv1.PersistentVolumeClaimSpec{
+			StorageClassName: &storageClass,
+			AccessModes:      []k8sv1.PersistentVolumeAccessMode{k8sv1.ReadWriteOnce},
+
```
Comment (Madhu-1) [inline e2e/common.go:40]: new line not required.

File: e2e/common.go [hunk 3]
```diff
@@ -1,12 +1,128 @@
...
+			VolumeMode: &volumemode,
+
```
Comment (Madhu-1) [inline e2e/common.go:42]: new line not required.

File: e2e/common.go [hunk 4]
```diff
@@ -1,12 +1,128 @@
...
+			Resources: k8sv1.ResourceRequirements{
+				Requests: k8sv1.ResourceList{
+					"storage": storageQuantity,
```
Comment (Madhu-1) [inline e2e/common.go:45]:
```suggestion
                                        v1.ResourceStorage: resource.MustParse(quantity),
```

File: e2e/common.go [hunk 5]
```diff
@@ -1,12 +1,128 @@
...
+	}
+	return pvc
+}
+
+// GetSamplePod returns a sample pod.
+func GetSamplePod(name string, pvcName string) *k8sv1.Pod {
+	pod := &k8sv1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Pod",
```
Comment (Madhu-1) [inline e2e/common.go:58]: same here as well.

File: e2e/common.go [hunk 6]
```diff
@@ -1,12 +1,128 @@
...
+	return pod
+}
+
+// WaitForPVCBound waits for a pvc with a given name and namespace to reach BOUND phase.
+func WaitForPVCBound(pvc *k8sv1.PersistentVolumeClaim, namespace string, pod *k8sv1.Pod) error {
+	pvc, err := DeployManagerObj.GetK8sClient().CoreV1().PersistentVolumeClaims(namespace).Create(context.TODO(), pvc, metav1.CreateOptions{})
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+	pod, err = DeployManagerObj.GetK8sClient().CoreV1().Pods(namespace).Create(context.TODO(), pod, metav1.CreateOptions{})
```
Comment (Madhu-1) [inline e2e/common.go:98]:
```suggestion
_, err = DeployManagerObj.GetK8sClient().CoreV1().Pods(namespace).Create(context.TODO(), pod, metav1.CreateOptions{})
```

File: e2e/common.go [hunk 7]
```diff
@@ -1,12 +1,128 @@
...
+
+	// Wait for namespace to terminate
+	err = utilwait.PollImmediate(interval, timeout, func() (done bool, err error) {
+		pvc, err = DeployManagerObj.GetK8sClient().CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(context.TODO(), pvc.Name, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			lastReason = fmt.Sprintf("error talking to k8s apiserver: %v", err)
+			return false, nil
+		}
+
+		if pvc.Status.Phase != k8sv1.ClaimBound {
```
Comment (Madhu-1) [inline e2e/common.go:115]: if the error is not found you need to continue of else this case cause panic as the Status might be nil

File: e2e/lvm/lvm_suite_test.go [hunk 8]
```diff
@@ -33,3 +38,61 @@ var _ = Describe("Validation test", func() {
...
+
+		Context("create pvc and pod", func() {
+			It("and verify bound status", func() {
+				By("Should be bound")
+				err := tests.WaitForPVCBound(filepvc, namespace, filepod)
+				fmt.Printf("%v", err)
+				Expect(err).To(BeNil())
+
+				err = tests.WaitForPVCBound(blockpvc, namespace, blockpod)
+				fmt.Printf("%v", err)
```
Comment (Madhu-1) [inline e2e/lvm/lvm_suite_test.go:91]: not required

File: e2e/lvm/lvm_suite_test.go [hunk 9]
```diff
@@ -33,3 +38,61 @@ var _ = Describe("Validation test", func() {
...
+				Expect(err).To(BeNil())
+			}
+
+		})
+
+		Context("create pvc and pod", func() {
+			It("and verify bound status", func() {
+				By("Should be bound")
+				err := tests.WaitForPVCBound(filepvc, namespace, filepod)
+				fmt.Printf("%v", err)
```
Comment (Madhu-1) [inline e2e/lvm/lvm_suite_test.go:87]: printf not required.

File: e2e/lvm/lvm_suite_test.go [hunk 10]
```diff
@@ -33,3 +38,61 @@ var _ = Describe("Validation test", func() {
...
+
+		AfterEach(func() {
+			err := tests.DeployManagerObj.GetK8sClient().CoreV1().Pods(namespace).Delete(context.TODO(), filepod.Name, metav1.DeleteOptions{})
+			if err != nil && !errors.IsNotFound(err) {
+				Expect(err).To(BeNil())
+			}
+			err = tests.DeployManagerObj.GetK8sClient().CoreV1().PersistentVolumeClaims(namespace).Delete(context.TODO(), filepvc.Name, metav1.DeleteOptions{})
+			if err != nil && !errors.IsNotFound(err) {
+				Expect(err).To(BeNil())
+			}
```
Comment (Madhu-1) [inline e2e/lvm/lvm_suite_test.go:70]: PVC deletion will not ensure PV is deleted. you might need to add a check to validate that PV is also deleted.

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: e2e/lvm/lvm_suite_test.go (hunk 10 above)
Comment (nbalacha) [reply e2e/lvm/lvm_suite_test.go:70]: I think we can take that up in a later PR. For now,  the LVMCluster cleanup will fail if the PV still exists.

File: e2e/common.go (hunk 7 above)
Comment (nbalacha) [reply e2e/common.go:115]: This should never be NotFound as the PVC was just created.

Comment (Madhu-1) [reply e2e/common.go:115]: :+1: above notFound error check need to be removed.

**Comment by riya-singhal31** [comment]
Updated the PR to address suggestions, thanks.

File: e2e/common.go [hunk 11]
```diff
@@ -1,12 +1,116 @@
...
 
 //nolint:errcheck
 func debug(msg string, args ...interface{}) {
 	ginkgo.GinkgoWriter.Write([]byte(fmt.Sprintf(msg, args...)))
 }
+
+// GetSamplePVC returns a sample pvc.
+func GetSamplePVC(storageClass string, quantity string, name string, volumemode k8sv1.PersistentVolumeMode) *k8sv1.PersistentVolumeClaim {
+	_, err := resource.ParseQuantity(quantity)
+	gomega.Expect(err).To(gomega.BeNil())
```
Comment (Madhu-1) [inline e2e/common.go:25]: this is not required.  as we are directly using quantity at line 38

File: e2e/common.go [hunk 12]
```diff
@@ -1,12 +1,116 @@
...
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
 )
 
 //nolint:errcheck
 func debug(msg string, args ...interface{}) {
 	ginkgo.GinkgoWriter.Write([]byte(fmt.Sprintf(msg, args...)))
 }
+
+// GetSamplePVC returns a sample pvc.
+func GetSamplePVC(storageClass string, quantity string, name string, volumemode k8sv1.PersistentVolumeMode) *k8sv1.PersistentVolumeClaim {
```
Comment (Madhu-1) [inline e2e/common.go:23]:
```suggestion
func GetSamplePVC(storageClass , quantity , name string, volumemode k8sv1.PersistentVolumeMode) *k8sv1.PersistentVolumeClaim {
```

File: e2e/common.go [hunk 13]
```diff
@@ -1,12 +1,116 @@
...
+					k8sv1.ResourceStorage: resource.MustParse(quantity),
+				},
+			},
+		},
+	}
+	return pvc
+}
+
+// GetSamplePod returns a sample pod.
+func GetSamplePod(name string, pvcName string) *k8sv1.Pod {
```
Comment (Madhu-1) [inline e2e/common.go:47]:
```suggestion
func GetSamplePod(name , pvcName string) *k8sv1.Pod {
```

File: e2e/common.go [hunk 14]
```diff
@@ -1,12 +1,116 @@
...
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("%v: %s", err, lastReason)
+	}
+
```
Comment (Madhu-1) [inline e2e/common.go:114]: You also need to check pod is in a running the state?

**Review by Madhu-1 — COMMENTED** [review]
small nits.

File: e2e/lvm/lvm_suite_test.go [hunk 15]
```diff
@@ -33,3 +36,59 @@ var _ = Describe("Validation test", func() {
...
+		BeforeEach(func() {
+			namespace = tests.TestNamespace
+			filepvc = tests.GetSamplePVC(tests.StorageClass, "5Gi", "lvmfilepvc", k8sv1.PersistentVolumeFilesystem)
+			filepod = tests.GetSamplePod("lvmfilepod", "lvmfilepvc")
+			blockpvc = tests.GetSamplePVC(tests.StorageClass, "5Gi", "lvmblockpvc", k8sv1.PersistentVolumeBlock)
+			blockpod = tests.GetSamplePod("lvmblockpod", "lvmblockpvc")
+		})
+
+		AfterEach(func() {
+			err := tests.DeployManagerObj.GetK8sClient().CoreV1().Pods(namespace).Delete(context.TODO(), filepod.Name, metav1.DeleteOptions{})
```
Comment (leelavg) [inline e2e/lvm/lvm_suite_test.go:61]: - if we are going to use `k8sClient` multiple times in multiple places why not register that in `BeforeSuite` (create a `var` at pkg level) if `SuiteFailed` to `False` and use that throughout?
- iirc there was also an effort to unify all clients used in the test and having that at pkg level might be helpful?

File: e2e/common.go (hunk 14 above)
Comment (riya-singhal31) [reply e2e/common.go:111]: Added the function for same.

File: e2e/common.go [hunk 16]
```diff
@@ -1,12 +1,116 @@
...
+				},
+			},
+		},
+	}
+	return pod
+}
+
+// WaitForPVCBound waits for a pvc with a given name and namespace to reach BOUND phase.
+func WaitForPVCBound(pvc *k8sv1.PersistentVolumeClaim, namespace string, pod *k8sv1.Pod) error {
+	pvc, err := DeployManagerObj.GetK8sClient().CoreV1().PersistentVolumeClaims(namespace).Create(context.TODO(), pvc, metav1.CreateOptions{})
```
[the remark this answers is not in the document]
Comment (riya-singhal31) [reply e2e/common.go:22]: Yes, thanks for the suggestion! 
That will need to be done in pkg/deploymanager. 
Will add this while changing multiple clients to single client and will create a separate PR for that concern.

File: e2e/lvm/lvm_suite_test.go (hunk 15 above)
Comment (riya-singhal31) [reply e2e/lvm/lvm_suite_test.go:61]: Yes, will unify all clients in the next PR.

File: e2e/common.go [hunk 17]
```diff
@@ -1,12 +1,116 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
 )
 
 //nolint:errcheck
 func debug(msg string, args ...interface{}) {
 	ginkgo.GinkgoWriter.Write([]byte(fmt.Sprintf(msg, args...)))
 }
+
+// GetSamplePVC returns a sample pvc.
```
[the remark this answers is not in the document]
Comment (riya-singhal31) [reply e2e/common.go]: Thanks, Updated.

File: e2e/lvm/lvm_suite_test.go [hunk 18]
```diff
@@ -33,3 +36,59 @@ var _ = Describe("Validation test", func() {
 		})
 	})
 })
+
+// Test to verify the PVC status
+var _ = Describe("PVC Status check", VerifyPVCStatus)
```
[the remark this answers is not in the document]
Comment (riya-singhal31) [reply e2e/lvm/lvm_suite_test.go]: Thanks for pointing this out Leela, 
I've now made separate files for separate tests now, as lvm_suite_test.go will makes more sense if it contains only setup and teardown part.

**Review by leelavg — APPROVED** [review]
- some more refactoring can be done however this pr is good enough for the functionality it's testing

**Review by nbalacha — APPROVED** [review]

**Pull request merged**